### PR TITLE
Add order property enforcement in memo optimization

### PR DESCRIPTION
**Pull request by arzuschen**
Updated from #158, complexity of implementation is greatly reduced.

Property enforcement is added without property propagation step. The cost calculation step is rewritten into a generalized function, without the separation between leaf group (and solver optimized group) and non-leaf group. Property requirement is added into the calculation of cost.

### CODE REVIEW

File: test/UnitTest.cs [hunk 1]
```diff
@@ -653,6 +653,47 @@ public void TestMemo()
...
+            memo.CalcStats(out int tlogics, out int tphysics);
+            Assert.AreEqual(8, memo.cgroups_.Count);
+            Assert.AreEqual(16, tlogics); Assert.AreEqual(17, tphysics);
+            Assert.AreEqual("4,1;6,4", string.Join(";", result));
+            var mstr = stmt.optimizer_.PrintMemo();
+            Assert.IsTrue(mstr.Contains("Summary: 16,17"));
+
+            sql = "select a1 from a, b where a1 <= b1 group by a1 order by a1";
+            result = TU.ExecuteSQL(sql, out stmt, out phyplan, option);
+            memo = stmt.optimizer_.memoset_[0];
```
Comment (zhouqingqing) [inline test/UnitTest.cs:686]: Can you find a way to assert we can push sort under NLJ?

Comment (arzuschen) [reply test/UnitTest.cs:686]: that should be doable

File: test/UnitTest.cs [hunk 2]
```diff
@@ -119,6 +119,17 @@ public static void CountOccurrences(string text, string pattern, int expected)
             var count = CountStr(text, pattern);
             Assert.AreEqual(expected, count);
         }
+        public static bool OccurenceWithOrder(string text, List<string> patternlist)
```
Comment (zhouqingqing) [inline test/UnitTest.cs:122]: we shall assert that the order is underneath join? try to check the plan (not the plan string) directly. - you can extend the ExecSQL() interface if we can't get the plan.

File: qpmodel/PhysicalNode.cs [hunk 3]
```diff
@@ -190,6 +190,10 @@ public ulong InclusiveMemory()
 
             return memory;
         }
+        public virtual PhysicProperty RequiredProperty() => null;
```
Comment (arzuschen) [inline qpmodel/PhysicalNode.cs:193]: add comment here

File: qpmodel/Program.cs [hunk 4]
```diff
@@ -189,6 +189,7 @@ static void Main(string[] args)
             a.queryOpt_.explain_.show_id_ = true;
             a.queryOpt_.explain_.show_estCost_ = a.queryOpt_.optimize_.use_memo_;
             a.queryOpt_.explain_.mode_ = ExplainMode.full;
+            a.queryOpt_.optimize_.enable_streamagg_ = false;
```
Comment (arzuschen) [inline qpmodel/Program.cs:192]: remove this change

File: qpmodel/optimizer.cs [hunk 5]
```diff
@@ -50,10 +50,77 @@ public class Property { }
...
-        KeyValuePair<Expr, DistributionType> distribution_;
+        public List<KeyValuePair<Expr, bool>> ordering_ = new List<KeyValuePair<Expr, bool>>();
+        
+        public bool IsPropertySupplied(PhysicNode node)
+        {
+            if (this.Equals(node.SuppiedProperty()))
+                return true;
+            return false;
+        }
+        // check if the left key of NLJoin match the property
```
Comment (arzuschen) [inline qpmodel/optimizer.cs:61]: move this part to nljoin

File: qpmodel/optimizer.cs [hunk 6]
```diff
@@ -296,6 +368,17 @@ public string Print()
             CountMembers(out int clogics, out int cphysics);
             var str = $"{clogics}, {cphysics}, [{logicSign_}][{minIncCost_}]: ";
             str += string.Join(",", exprList_);
+
+            // add property optimal member
+            if (minMember_.Count > 0)
+            {
+                List<string> l = new List<string>();
+                foreach (var pair in minMember_)
+                    l.Add($"property:{pair.Key.ToString()}, member:{pair.Value.Key}, cost:{pair.Value.Value.ToString("0.##")}");
```
Comment (arzuschen) [inline qpmodel/optimizer.cs:377]: remove the first ToString (unnecessary)

File: qpmodel/optimizer.cs [hunk 7]
```diff
@@ -342,102 +425,135 @@ public void ExploreGroup(Memo memo)
             explored_ = true;
         }
 
-        // propagate the property requirement from the top
-        public void PropagateProperty(PhysicProperty property)
+        // preprocess exprList_ to get whether property can be supplied and subproperties
+        // also find members that can propagate required property and split them
```
Comment (arzuschen) [inline qpmodel/optimizer.cs:429]: add more comments about the input output lists

File: qpmodel/optimizer.cs [hunk 8]
```diff
@@ -342,102 +425,135 @@ public void ExploreGroup(Memo memo)
...
+                    fulfilled.Add(property?.IsPropertySupplied(member.physic_) ?? true);
+                    subproplist.Add(subproperties);
                 }
             }
-
-            minIncCost_ = mincost;
-            Debug.Assert(IsSolverOptimizedGroup() ||
-                minMember_.physic_.InclusiveCost() == minIncCost_);
-            return minMember_;
+            Debug.Assert(candidates.Count == fulfilled.Count && candidates.Count == subproplist.Count);
```
Comment (arzuschen) [inline qpmodel/optimizer.cs:458]: add assertion here:
if property is null, candidate number = physic member number

File: qpmodel/optimizer.cs (hunk 7 above)
Comment (arzuschen) [reply qpmodel/optimizer.cs:429]: change to tuple

File: qpmodel/optimizer.cs [hunk 9]
```diff
@@ -342,102 +425,135 @@ public void ExploreGroup(Memo memo)
...
-            double mincost = double.MaxValue;
-            foreach (var v in exprList_)
+            foreach (var member in exprList_)
             {
-                var physic = v.physic_;
-                if (physic != null && physic.Cost() < mincost)
+                if (member.physic_ is null) continue;
+
+                var subproperties = ProcessChildProperty(member.physic_, property, out bool propagated);
+                if (propagated)
```
Comment (arzuschen) [inline qpmodel/optimizer.cs:442]: add example here

File: qpmodel/optimizer.cs [hunk 10]
```diff
@@ -342,102 +425,135 @@ public void ExploreGroup(Memo memo)
...
+                // when no required property and propagated is not all null
+                // meaning certain property is propagated to child
+                // the status is recorded
+                foreach (var prop in childprop)
+                    if (prop != null) propagated = true;
+            }
+            return childprop;
+        }
+
+        internal CGroupMember CalculateMinInclusiveCostMember(PhysicProperty property = null)
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:484]: let's try to follow the original code path per discussion.

**Pull request closed without merging**